File: 10.Math数学/leetcode_1622_Fancy.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Fancy {
private:
    vector<int> lst;
    long long mulNum, addNum, modNum;

    long long quickPow(long long a, int b){
        long long res = 1;
        while (b){
            if (b&1){
                res = (res*a)%modNum;
            }
            a = (a*a)%modNum;
            b >>= 1;
        }
        return res;
    }
public:
    Fancy() {
        mulNum = 1ll;
        addNum = 0ll;
        modNum = (long long)1e9+7;
    }
    
    void append(int val) {
        lst.push_back(((val-addNum+modNum)*quickPow(mulNum, modNum-2))%modNum);
    }
    
    void addAll(int inc) {
        addNum += inc;
    }
    
    void multAll(int m) {
        addNum = (addNum*m)%modNum;
        mulNum = (mulNum*m)%modNum;
    }
    
    int getIndex(int idx) {
        if (idx >= lst.size()){
            return -1;
        }
        return ((1ll*lst[idx]*mulNum)%modNum+addNum)%modNum;
    }
};

int main(){
    Fancy f;
    vector<string> ops = {"append","addAll","append","multAll","getIndex","addAll","append","multAll","getIndex","getIndex","getIndex"};
    vector<int> nums = {2,3,7,2,0,3,10,2,0,1,2};
    for (int i=0; i<ops.size(); i++){
        if (ops[i] == "append"){
            f.append(nums[i]);
        }
        else if (ops[i] == "addAll"){
            f.addAll(nums[i]);
        }
        else if (ops[i] == "multAll"){
            f.multAll(nums[i]);
        }
        else{
            cout << f.getIndex(nums[i]) << endl;
        }
    }
}

File: 10.Math数学/leetcode_1622_Fancy.py
modNum = 1_000_000_007
class Fancy:

    def __init__(self):
        self.lst = []
        self.addNum = 0
        self.mulNum = 1

    def append(self, val: int) -> None:
        self.lst.append(((val-self.addNum)*pow(self.mulNum, -1, modNum))%modNum)

    def addAll(self, inc: int) -> None:
        self.addNum += inc

    def multAll(self, m: int) -> None:
        self.mulNum = (self.mulNum*m)%modNum
        self.addNum = (self.addNum*m)%modNum

    def getIndex(self, idx: int) -> int:
        if idx >= len(self.lst):
            return -1
        return (self.lst[idx]*self.mulNum + self.addNum)%modNum

if __name__ == "__main__":
    f = Fancy()
    ops = ["append","addAll","append","multAll","getIndex","addAll","append","multAll","getIndex","getIndex","getIndex"]
    nums = [[2],[3],[7],[2],[0],[3],[10],[2],[0],[1],[2]]
    for i in range(len(ops)):
        if ops[i] == "append":
            f.append(nums[i][0])
        elif ops[i] == "addAll":
            f.addAll(nums[i][0])
        elif ops[i] == "multAll":
            f.multAll(nums[i][0])
        else:
            print(f.getIndex(nums[i][0]))
